libs/pilight/core/common.c

```diff
@@ -1019,3 +1024,161 @@ int file_get_contents(char *file, char **content) {
 	fclose(fp);
 	return 0;
 }
+
+int checkdnsrr(const char *domain, const char *type)
+{
+	logprintf(LOG_STACK, "%s(%s,%s)", __FUNCTION__, domain ? domain : "(NULL)", type ? type : "(NULL)");
+	return 0;	// NYI.
+}
+
+#if defined(__FreeBSD__) || defined(_WIN32)
+#define use_regexp 0
+#else
+#define use_regexp 1
+#endif
+
+// Do NOT write the next as range A-Za-z0-9 because the string may be used without RE.
+#define AtoZ_atoz_0to9 "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
+#define white_spaces " \t"
+
+// Next email_local_part_valid_chars is defined in a way that it can be used without RE
+// and within [] of a posix extended regular expression. Therefore the dash comes first
+// because at later position it would mean range (and it can't be escaped because
+// a \ inside [] is literal - this is different from perl RE).
+#define email_local_part_valid_chars "-" AtoZ_atoz_0to9 ".!#$%&'*+\\/=?^_`{|}~"
+
+#if use_regexp
+// The RE to check email addr is compiled only once and used commonly.
+
+static regex_t check_email_common_RE = { 0 };
+static const regex_t * volatile check_email_compiled_RE = NULL;
+
+static void check_email_free_RE() {
+	if(check_email_compiled_RE != NULL) {	// protect against unexpected calls.
+		regfree(&check_email_common_RE);
+		check_email_compiled_RE = NULL;
+	}
+}
+#endif
+
+/**
+ * Check if an email address is valid.
+ * @param char* addr The email address to check. Leading+trailing white spaces are accepted.
+ * @param bool allow_lists If true, addr is allowed to be a comma separated list of email addresses.
+ * @param bool check_domain_can_mail If true, check if domain exists in DNS system and accepts email.
+ * @return int Values < 0 indicate error. Values >= 0 indicate success.
+ * Where -1: syntax error, -2: DNS check for a domain failed, -8: severe internal error.
+ */
+int check_email_addr(const char *addr, int allow_lists, int check_domain_can_mail) {
+	if(addr == NULL) {
+		return -1;
+	}
+#if use_regexp
+	const regex_t *RE = NULL;
+	int reti = 0;
+
+	static const char email_pattern[] =
+		"^[" email_local_part_valid_chars "]+"
```

Comment:
because email_local_part_valid_chars is used more than once in the code (in the use_regex part and the not use_regex part). When integrating the expression directly, there would be two places where this set of chars is written. If you change one it tends to forget the other. In particular because it is not even clear that it should be the same set of chars. Using a name for the set and define it at one place makes all this clear.